common/certificate/index.ts

```diff
@@ -0,0 +1,337 @@
+    }
+
+    if (signaturePupil) {
+        certificate.signaturePupil = Buffer.from(signaturePupil, "utf-8");
+    }
+
+    certificate.signatureDate = new Date();
+    certificate.signatureLocation = signatureLocation;
+    certificate.state = "approved";
+
+    await getManager().save(ParticipationCertificate, certificate);
+
+    const rendered = await createPDFBinary(certificate, getCertificateLink(certificate, "de"), "de");
+
+    const certificateLink = createAutoLoginLink(certificate.student, `/settings`);
+    const mail = mailjetTemplates.CERTIFICATESIGNED({
+        certificateLink,
+        pupilFirstname: certificate.pupil.firstname,
+        studentFirstname: certificate.student.firstname
+    }, rendered.toString("base64"));
+    await sendTemplateMail(mail, certificate.student.email);
+    await Notification.actionTaken(certificate.student, "student_certificate_sign", {
+        uniqueId: `${certificate.id}`,
+        certificateLink,
+        pupil: certificate.pupil
+    });
+
+}
+
+/* ------------------------- INTERNAL CERTIFICATE UTILITIES --------------------------------------- */
+
+const _templates: { [name: string]: { [key in Language | "default"]?: EJS.ClientFunction } } = {};
+
+/* Loads the template from the /assets folder, falls back to the default language if fallback is true */
+function loadTemplate(name, lang: Language, fallback: boolean = true): EJS.ClientFunction {
+    if (_templates[name] && _templates[name][lang]) {
+        return _templates[name][lang];
+    }
+
+    let file = path.join(ASSETS, `${name}.${lang}.html`);
+    console.log("Loading template from ", file);
+
+    if (existsSync(file)) {
+        const result = readFileSync(file, "utf8");
+        if (!_templates[name]) {
+            _templates[name] = {};
+        }
+
+        const compiled = EJS.compile(result);
+
+        _templates[name][lang] = compiled;
+        return compiled;
+    } else {
+        if (!fallback || lang === DefaultLanguage) {
+            throw new Error(`Cannot find template '${file}`);
+        }
+
+        return loadTemplate(name, DefaultLanguage, /*fallback:*/ false);
+    }
+}
+
+/* Map the certificate data to something the frontend can work with while keeping user data secret */
+function exposeCertificate({ student, pupil, state, ...cert }: ParticipationCertificate, to: Student | Pupil): IExposedCertificate {
+    return {
+        ...cert,
+        // NOTE: user.id is NOT unique, as Students and Pupils can have the same id
+        userIs: pupil.wix_id === to.wix_id ? "pupil" : "student",
+        pupil: { firstname: pupil.firstname, lastname: pupil.lastname },
+        student: { firstname: student.firstname, lastname: student.lastname },
+        state: state as CertificateState
+    };
+}
+
+function getCertificateLink(certificate: ParticipationCertificate, lang: Language) {
+    return "http://verify.corona-school.de/" + certificate.uuid + "?lang=" + lang;
```

Comment:
Is this link still up to date?

Reply:
Yes, see e.g. http://verify.corona-school.de/81C18DEC40?lang=de. And we can't change it, as these links are generated into the certificate PDFs, so there are already hundreds of these links out there. But fortunately they were smart enough back then to move these under their own subdomain, so that the rest of the landscape can move independently ...